### ISSUE TITLE
Async API

### ISSUE THREAD

**Issue by Kixunil** [laanwj/cln4rust#1]
Consider providing async API via `tokio`/`futures`.

**Comment by laanwj** [comment]
Yes that would be useful.

**Comment by Kixunil** [comment]
What do you think would be the best approach here?

* Make `futures` and `tokio-io` a mandatory dependency, changing all sync implementations to async - the user has to wait for futures
* Make `futures` and `tokio-io` mandatory dependency, implement a new async API, implement sync API using async internally to de-duplicate the code.
* Make `futures` and `tokio-io` optional, providing async API behind `async` feature. The code might get duplicated, but I guess some part could be common.

**Comment by laanwj** [comment]
It seems overkill to require it as a mandatory dependency, or to require the user to use it. Many uses will involve simple single-threaded use, where the current function-call emulation is very convenient.
(it also mimics the Python binding)

So I guess the third approach is best. But to be honest, I've never done anything with rust and async APIs so I don't even know what this would look like in this case.

**Comment by Kixunil** [comment]
Yeah, it makes sense. I did some async stuff, so I'm going to help with this. I think having a separate struct `async::LightningRPC` will be clearly better than bothering async users with typing `async_` in front of every method, so I'll start with this approach. I'll try to de-duplicate as much as reasonable.

I will use "impl Trait" feature, which requires quite new compiler, but that won't be a problem thanks to feature flag. (Those attempting to use async stuff with old compiler are complicating their lives anyway.) We can change this in backward-compatible way any time and maybe we will have to eventually because of some limitations of `impl Trait`.

**Comment by laanwj** [comment]
Thanks for working on this !

Yes, I think it's acceptable to have higher compiler requirement for the async API, being an experimental thing in a still-experimental crate.

> I think having a separate struct async::LightningRPC will be clearly better than bothering async users with typing async_ in front 

Agree. Also it's good to have more or less equivalence between LightningRPC methods and actual RPC calls, for documentation purposes.

So yes, just cloning the high-level API with different signatures (to return some Async-handle thing) probably makes sense.

Apart from `pay`, which due to the huge number of optional arguments is handled differently, and doc comments, it's boilerplate that could potentially be generated from the structures in `requests`, btw.

**Comment by Kixunil** [comment]
I've been looking into this and there's one problem with async right now: we don't know beforehand how many bytes will the response take and `strason` doesn't support feeding bytes gradually into the parser.

So far I see these options:

* Patch strason to enable parsing asynchronous streams
* Patch strason to have `.is_eof()` method on its error type and then just attempt to deserialize it until it succeeds. This is the simplest but my intuition says the performance would be horrible.
* Use serde_json (which has `.is_eof()` method on error type) to scan the stream. Requires no patching, but the performance is even worse
* Create more optimized scanner - but if we're doing that, why not just create async parser?

As much as I'd love adding async capability to strason, I figured it's too complex for me to do it in my free time right now, sorry. Do you have any idea to tackle this problem?

**Comment by vincenzopalazzo** [comment]
I am working on the official 0.3 of this library finally, so I spend some time of this and with the current rust this should be a lot of easier https://github.com/laanwj/cln4rust/pull/141

Thanks for chatting about this, and keep this issue open to allow me to keep tracking it!
